README.rst
===============================
py65asm
===============================

.. image:: https://badge.fury.io/py/py65asm.png
    :target: http://badge.fury.io/py/py65asm
    
.. image:: https://travis-ci.org/docmarionum1/py65asm.png?branch=master
        :target: https://travis-ci.org/docmarionum1/py65asm

.. image:: https://pypip.in/d/py65asm/badge.png
        :target: https://crate.io/packages/py65asm?version=latest


Python 6502 Assembler

* Free software: BSD license
* Documentation: http://py65asm.rtfd.org.

Features
--------

* TODO

HISTORY.rst
.. :changelog:

History
-------

0.0.0 (2014-02-19)
++++++++++++++++++

* First release on PyPI.

tests/files/test.asm
var = $99

		.org $C000
start	lda #$ff
		jmp start

		.org $fffd
		.byte var
		.word start


tests/test_py65asm.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_py65asm
----------------------------------

Tests for `py65asm` module.
"""

import unittest

from py65asm import py65asm


class TestPy65asm(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()

tests/test_assembler.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_assembler
----------------------------------

Tests for `py65asm` module.
"""

import unittest, os

from py65asm.assembler import Assembler


class TestAssembler(unittest.TestCase):

    def _asm(self):
        return Assembler()

    def setUp(self):
        pass

    def test_im_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x55
        self.assertEqual(a.getArgument("#$FF"), ('im', 0xff))
        self.assertEqual(a.getArgument("#%10101011"), ('im', 0b10101011))
        self.assertEqual(a.getArgument("#010"), ('im', 010))
        self.assertEqual(a.getArgument("#123"), ('im', 123))
        self.assertEqual(a.getArgument("#var"), ('im', a.symbols['var']))

    def test_z_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x55
        self.assertEqual(a.getArgument("$FF"), ('z', 0xff))
        self.assertEqual(a.getArgument("%10101011"), ('z', 0b10101011))
        self.assertEqual(a.getArgument("010"), ('z', 010))
        self.assertEqual(a.getArgument("123"), ('z', 123))
        self.assertEqual(a.getArgument("var"), ('z', a.symbols['var']))

    def test_zx_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x55
        self.assertEqual(a.getArgument("$FF,X"), ('zx', 0xff))
        self.assertEqual(a.getArgument("%10101011,X"), ('zx', 0b10101011))
        self.assertEqual(a.getArgument("010,X"), ('zx', 010))
        self.assertEqual(a.getArgument("123,X"), ('zx', 123))
        self.assertEqual(a.getArgument("var,X"), ('zx', a.symbols['var']))

    def test_zy_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x55
        self.assertEqual(a.getArgument("$FF,Y"), ('zy', 0xff))
        self.assertEqual(a.getArgument("%10101011,Y"), ('zy', 0b10101011))
        self.assertEqual(a.getArgument("010,Y"), ('zy', 010))
        self.assertEqual(a.getArgument("123,Y"), ('zy', 123))
        self.assertEqual(a.getArgument("var,Y"), ('zy', a.symbols['var']))

    def test_a_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x555
        self.assertEqual(a.getArgument("$FFFF"), ('a', 0xffff))
        self.assertEqual(a.getArgument("$00FF"), ('a', 0x00ff))
        self.assertEqual(a.getArgument("%1010101110101011"), ('a', 0b1010101110101011))
        self.assertEqual(a.getArgument("01110"), ('a', 01110))
        self.assertEqual(a.getArgument("1234"), ('a', 1234))
        self.assertEqual(a.getArgument("var"), ('a', a.symbols['var']))

    def test_ax_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x555
        self.assertEqual(a.getArgument("$FFFF,X"), ('ax', 0xffff))
        self.assertEqual(a.getArgument("%1010101110101011,X"), ('ax', 0b1010101110101011))
        self.assertEqual(a.getArgument("01110,X"), ('ax', 01110))
        self.assertEqual(a.getArgument("1234,X"), ('ax', 1234))
        self.assertEqual(a.getArgument("var,X"), ('ax', a.symbols['var']))

    def test_ay_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x555
        self.assertEqual(a.getArgument("$FFFF,Y"), ('ay', 0xffff))
        self.assertEqual(a.getArgument("%1010101110101011,Y"), ('ay', 0b1010101110101011))
        self.assertEqual(a.getArgument("01110,Y"), ('ay', 01110))
        self.assertEqual(a.getArgument("1234,Y"), ('ay', 1234))
        self.assertEqual(a.getArgument("var,Y"), ('ay', a.symbols['var']))

    def test_ix_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x55
        self.assertEqual(a.getArgument("($FF,X)"), ('ix', 0xff))
        self.assertEqual(a.getArgument("(%10101011,X)"), ('ix', 0b10101011))
        self.assertEqual(a.getArgument("(010,X)"), ('ix', 010))
        self.assertEqual(a.getArgument("(123,X)"), ('ix', 123))
        self.assertEqual(a.getArgument("(var,X)"), ('ix', a.symbols['var']))

    def test_iy_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x55
        self.assertEqual(a.getArgument("($FF),Y"), ('iy', 0xff))
        self.assertEqual(a.getArgument("(%10101011),Y"), ('iy', 0b10101011))
        self.assertEqual(a.getArgument("(010),Y"), ('iy', 010))
        self.assertEqual(a.getArgument("(123),Y"), ('iy', 123))
        self.assertEqual(a.getArgument("(var),Y"), ('iy', a.symbols['var']))

    def test_i_argument(self):
        a = self._asm()
        a.symbols['var'] = 0x555
        self.assertEqual(a.getArgument("($FFFF)"), ('i', 0xffff))
        self.assertEqual(a.getArgument("(%1010101110101011)"), ('i', 0b1010101110101011))
        self.assertEqual(a.getArgument("(01110)"), ('i', 01110))
        self.assertEqual(a.getArgument("(1234)"), ('i', 1234))
        self.assertEqual(a.getArgument("(var)"), ('i', a.symbols['var']))

    def test_A_argument(self):
        a = self._asm()
        self.assertEqual(a.getArgument("A"), ('im', 'A'))

    def test_assemble_line(self):
        a = self._asm()
        self.assertEqual(a.assemble("LDA #55"), [169, 55])
        self.assertEqual(a.assemble("LDA #$55"), [169, 0x55])
        self.assertEqual(a.assemble("LDA $55"), [165, 0x55])
        self.assertEqual(a.assemble("LDA $555"), [173, 0x55, 0x5])

    def test_assemble_lines(self):
        a = self._asm()
        self.assertEqual(a.assemble("LDA #55\nSBC $33,X"), [169, 55, 245, 0x33])

    def test_labels(self):
        a = self._asm()
        self.assertEqual(a.assemble("loop: DEX\nJMP loop"), [202, 76, 0, 0])
        self.assertEqual(
            a.assemble("JMP loop\nDEX\nDEX\nDEX\nloop"), 
            [76, 06, 00, 202, 202, 202]
        )
        self.assertEqual(
            a.assemble("label1: JMP label1\nJMP label2\nlabel2: JMP label1\nJMP label2"),
            [76, 00, 00, 76, 06, 00, 76, 00, 00, 76, 06, 00]
        )

    def test_branch(self):
        a = self._asm()
        self.assertEqual(a.assemble("label: DEX\nDEX\nBNE label"),
            [202, 202, 208, 252]
        )
        self.assertEqual(a.assemble("BNE label\nDEX\nDEX\nlabel"),
            [208, 2, 202, 202]
        )

    def test_variables(self):
        a = self._asm()
        self.assertEqual(a.assemble("var = $ff\nlda #var"),[169, 255])
        self.assertEqual(a.assemble("var = $ff\nlda var"),[165, 255])
        self.assertEqual(a.assemble("var = $ffff\nlda var"),[173, 255, 255])
        self.assertEqual(
            a.assemble("var1 = *\nlda #var1\nvar2 = *\nlda #var2"),
            [169, 0, 169, 2]
        )

    def test_byte(self):
        a = self._asm()
        self.assertEqual(a.assemble(".BYTE $0A"), [0xa])

    def test_word(self):
        a = self._asm()
        self.assertEqual(a.assemble(".WORD $0A"), [0xa, 0x00])
        self.assertEqual(a.assemble(".WORD $AAA"), [0xaa, 0xa])

    def test_org(self):
        a = self._asm()
        self.assertEqual(
            a.assemble(".BYTE 01\n.ORG 10\n.BYTE 02"), 
            [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2]
        )
        self.assertEqual(
            a.assemble(".ORG $8000\n.BYTE 1\nlabel: .BYTE 2\nJMP label"), 
            [1, 2, 76, 01, 0x80]
        )
        self.assertEqual(
            a.assemble("a = 5\n.ORG $8000\n.BYTE 1\nlabel: .BYTE 2\nJMP label"), 
            [1, 2, 76, 01, 0x80]
        )
        self.assertEqual(
            a.assemble(".ORG $8000\n.BYTE 1\n.ORG $8005\nlabel: .BYTE 2\nJMP label"), 
            [1, 0, 0, 0, 0, 2, 76, 01, 0x80]
        )

    def test_assemble_file(self):
        a = self._asm()

        f = os.path.join(
            os.path.dirname(os.path.realpath(__file__)), 
            "files", "test.asm"
        )

        out = a.assemble(open(f))

        self.assertEqual(len(out), 0x4000)
        self.assertEqual(out[:5], [0xa9, 0xff, 0x4c, 0x00, 0xc0])
        self.assertEqual(out[-3:], [0x99, 0x00, 0xc0])

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()

py65asm/assembler.py
from ops import ops
import re

num_formats = [
    '\$([\dabcdefABCDEF]{1,4})',
    '%([01]{1,16})',
    '0([01234567]{1,6})',
    '(\d{1,5})',
    '(\w[\w\d]*)'
]

num_bases = [16, 2, 8, 10, None]

arg_regex = [
    ('im', ''.join([
        '^#', '(?:', '|'.join(num_formats), ')$'
    ])),
    ('z', ''.join([
        '^(?:', '|'.join(num_formats), ')$'
    ])),
    ('zx', ''.join([
        '^(?:', '|'.join(num_formats), ')', ',X$'
    ])),
    ('zy', ''.join([
        '^(?:', '|'.join(num_formats), ')', ',Y$'
    ])),
    ('ix', ''.join([
        '^\(', '(?:',
        '|'.join(num_formats),
        ')', ',X\)$',
    ])),
    ('iy', ''.join([
        '^\(', '(?:',
        '|'.join(num_formats),
        ')', '\),Y$',
    ])),
    ('i', ''.join([
        '^\(', '(?:',
        '|'.join(num_formats),
        ')', '\)$',
    ])),
]


class Assembler:

    def __init__(self, org=None):
        self.symbols = {}
        self.out = []
        if org is not None:
            self.org = org
            self.start = self.org
        else:
            self.org = 0
            self.start = self.org

    def assemble(self, asm, output_dest=None):
        self.out = []

        if type(asm) == str:
            lines = asm.split('\n')
        elif type(asm) == file:
            lines = list(asm.readlines())

        for l in lines:
            l = l.strip()
            if l:
                if l[0] == ";": #Comment
                    continue

                if ";" in l: #Strip trailing comment
                    l = l[:l.index(";")]

                tokens = l.split()

                self.assembleTokens(tokens)

        self.resolveLabels()

            
        return self.out

    def assembleTokens(self, tokens):
        op = tokens[0].upper()
        if op in ops:
            if len(tokens) == 1: #Implied
                self.out.append(ops[op]['im'])
                return

            t, n = self.getArgument(tokens[1])
            if n:
                # If zero page is not available switch to absolute
                if t not in ops[op] and t == 'z': 
                    t = 'a'

                self.out.append(ops[op][t])
                if t in ['a', 'ax', 'ay', 'i']:
                    self.out.append(n & 0xff)
                    self.out.append(n >> 8)
                else:
                    self.out.append(n)
            else: #Unresolved variable
                self.out.append(tokens[0])
                self.out.append(tokens[1])
        elif op == ".BYTE":
            n = self.getNumber(tokens[1])
            if n:
                self.out.append(n)
            else:
                self.out.append(tokens[1])
        elif op == ".WORD":
            n = self.getNumber(tokens[1])
            print n
            if n:
                self.out.append(n & 0xff)
                self.out.append(n >> 8)
            else:
                self.out.append("WORD")
                self.out.append(tokens[1])
        elif op == ".ORG":
            n = self.getNumber(tokens[1])
            #Is this before any code has been generated?  Set the start to this
            if len(self.out) == 0 or not [i for i in self.out if type(i) != str]:
                self.start = n
            else:
                #self.out.extend([0]**n - len)
                self.out.append("ORG")
                self.out.append(str(n))
        else: #Label or variable
            if len(tokens) > 1 and tokens[1] == "=": #variable
                if tokens[2] == "*": #label = * is equal to lable:
                    self.out.append("LABEL")
                    self.out.append(tokens[0])
                else:
                    n  = self.getNumber(tokens[2])
                    self.symbols[tokens[0]] = n
            else:
                self.out.append("LABEL")
                self.out.append(tokens[0].rstrip(":"))
                if len(tokens) > 1: #Label on same line as code
                    self.assembleTokens(tokens[1:])

    def resolveLabels(self):
        i = 0
        while i < len(self.out):

            if type(self.out[i]) == str and self.out[i] == "ORG":
                self.out = self.out[:i] + [0]*(int(self.out[i+1])-i - self.start) + self.out[i+2:]
            elif type(self.out[i]) == str and self.out[i] == "LABEL":
                self.out.pop(i) #Remove the "LABEL" tag
                self.symbols[self.out[i]] = i + self.start #Add the label to the symbols
                #Resolve this label
                j = 0
                l = len(self.out)

                while j < len(self.out):
                    if type(self.out[j]) == str and j != i and self.out[i] in self.out[j]:

                        if type(self.out[j-1]) == int: #BYTE
                            self.out[j] = self.getNumber(self.out[j])
                            continue
                        elif self.out[j-1] == "WORD":
                            print self.out[j-1], self.out[j], self.symbols
                            n = self.getNumber(self.out[j])
                            self.out[j-1] = n & 0xff
                            self.out[j] = n >> 8
                            continue


                        new = []
                        t, n = self.getArgument(self.out[j])
                        op = self.out[j-1].upper()
                        # If zero page is not available switch to absolute
                        if t not in ops[op] and t == 'z': 
                            t = 'a'

                        new.append(ops[op][t])

                        if op in ["BCC", "BCS", "BEQ", "BMI", "BNE", "BPL", "BVC"]:
                            if n < j:
                                d = n - j
                            else:
                                d = n - j - 1

                            if d > 127 or d < -128:
                                    raise Exception("Branch target too far")

                            new.append(d & 0xff)
                        elif t in ['a', 'ax', 'ay', 'i']:
                            if (j + self.start) < n:  #Adjust the labels position for the extra byte
                                self.symbols[self.out[i]] += 1
                                n += 1
                                i += 1
                            new.append(n & 0xff)
                            new.append(n >> 8)
                        else:
                            new.append(n)

                        self.out = self.out[:j-1] + new + self.out[j+1:]

                    j += 1
                self.out.pop(i)
                continue

            i += 1


    def getNumber(self, arg):
        for i in range(len(num_bases)):
            if num_bases[i]:
                s = re.match(num_formats[i], arg)
                if s and s.group(1):
                    return int(s.group(1), num_bases[i])
            else:
                return self.symbols.get(arg, None)


    def getArgument(self, arg):
        if re.match('A', arg):
            return ('im', 'A')

        for r in arg_regex:
            s = re.match(r[1], arg)
            if s:
                if s.group(1):
                    t, n, v  = (r[0], int(s.group(1), 16), s.group(1))
                    break
                elif s.group(2):
                    t, n, v  = (r[0], int(s.group(2), 2), s.group(2))
                    break
                elif s.group(3):    
                    t, n, v  = (r[0], int(s.group(3), 8), s.group(3))
                    break
                elif s.group(4):    
                    t, n, v  = (r[0], int(s.group(4), 10), s.group(4))
                    break
                elif s.group(5):    
                    t, n, v  = (r[0], self.symbols.get(s.group(5), None), s.group(5))
                    break


        if t[0] == "z" and (n > 0xff or len(v) == 4):
            t = t.replace("z", "a")

        return t, n



py65asm/ops.py
ops = {   
    'AAC': {'im': 11},
    'AAX': {'a': 143, 'ix': 131, 'z': 135, 'zy': 151},
    'ADC': {'a': 109, 'ax': 125, 'ay': 121, 'im': 105, 'ix': 97, 'iy': 113, 'z': 101, 'zx': 117},
    'AND': {'a': 45, 'ax': 61, 'ay': 57, 'im': 41, 'ix': 33, 'iy': 49, 'z': 37, 'zx': 53},
    'ARR': {'im': 107},
    'ASL': {'a': 14, 'ax': 30, 'im': 10, 'z': 6, 'zx': 22},
    'ASR': {'im': 75},
    'ATX': {'im': 171},
    'AXA': {'ay': 159, 'iy': 147},
    'AXS': {'im': 203},
    'BCC': {'z': 144},
    'BCS': {'z': 176},
    'BEQ': {'z': 240},
    'BMI': {'z': 48},
    'BNE': {'z': 208},
    'BPL': {'z': 16},
    'BVC': {'z': 112},
    'BIT': {'a': 44, 'z': 36},
    'BRK': {'im': 0},
    'CLC': {'im': 24},
    'CLD': {'im': 216}, 
    'CLI': {'im': 88}, 
    'CLV': {'im': 184},
    'CMP': {'a': 205, 'ax': 221, 'ay': 217, 'im': 201, 'ix': 193, 'iy': 209, 'z': 197, 'zx': 213},
    'CPX': {'a': 236, 'im': 224, 'z': 228},
    'CPY': {'a': 204, 'im': 192, 'z': 196},
    'DCP': {'a': 207, 'ax': 223, 'ay': 219, 'ix': 195, 'iy': 211, 'z': 199, 'zx': 215},
    'DEC': {'a': 206, 'ax': 222, 'z': 198, 'zx': 214},
    'DEX': {'im': 202},
    'DEY': {'im': 136},
    'EOR': {'a': 77, 'ax': 93, 'ay': 89, 'im': 73, 'ix': 65, 'iy': 81, 'z': 69, 'zx': 85},
    'INC': {'a': 238, 'ax': 254, 'z': 230, 'zx': 246},
    'INX': {'im': 232},
    'INY': {'im': 200},
    'ISC': {'a': 239, 'ax': 255, 'ay': 251, 'ix': 227, 'iy': 243, 'z': 231, 'zx': 247},
    'JMP': {'a': 76, 'i': 108},
    'JSR': {'a': 32},
    'KIL': {'im': 2},
    'LAR': {'ay': 187},
    'LAX': {'a': 175, 'ay': 191, 'ix': 163, 'iy': 179, 'z': 167, 'zy': 183},
    'LDA': {'a': 173, 'ax': 189, 'ay': 185, 'im': 169, 'ix': 161, 'iy': 177, 'z': 165, 'zx': 181},
    'LDX': {'a': 174, 'ay': 190, 'im': 162, 'z': 166, 'zy': 182},
    'LDY': {'a': 172, 'ax': 188, 'im': 160, 'z': 164, 'zx': 180},
    'LSR': {'a': 78, 'ax': 94, 'im': 74, 'z': 70, 'zx': 86},
    'NOP': {'im': 26},
    'ORA': {'a': 13, 'ax': 29, 'ay': 25, 'im': 9, 'ix': 1, 'iy': 17, 'z': 5, 'zx': 21},
    'PHA': {'im': 72}, 
    'PHP': {'im': 8}, 
    'PLA': {'im': 104}, 
    'PLP': {'im': 40},
    'RLA': {'a': 47, 'ax': 63, 'ay': 59, 'ix': 35, 'iy': 51, 'z': 39, 'zx': 55},
    'ROL': {'a': 46, 'ax': 62, 'im': 42, 'z': 38, 'zx': 54},
    'ROR': {'a': 110, 'ax': 126, 'im': 106, 'z': 102, 'zx': 118},
    'RRA': {'a': 111, 'ax': 127, 'ay': 123, 'ix': 99, 'iy': 115, 'z': 103, 'zx': 119},
    'RTI': {'im': 64},
    'RTS': {'im': 96},
    'SBC': {'a': 237, 'ax': 253, 'ay': 249, 'im': 233, 'ix': 225, 'iy': 241, 'z': 229, 'zx': 245},
    'SEC': {'im': 56}, 
    'SED': {'im': 248}, 
    'SEI': {'im': 120},
    'SLO': {'a': 15, 'ax': 31, 'ay': 27, 'ix': 3, 'iy': 19, 'z': 7, 'zx': 23},
    'SRE': {'a': 79, 'ax': 95, 'ay': 91, 'ix': 67, 'iy': 83, 'z': 71, 'zx': 87},
    'STA': {'a': 141, 'ax': 157, 'ay': 153, 'ix': 129, 'iy': 145, 'z': 133, 'zx': 149},
    'STX': {'a': 142, 'z': 134, 'zy': 150},
    'STY': {'a': 140, 'z': 132, 'zx': 148},
    'SXA': {'ay': 158},
    'SYA': {'ax': 156},
    'TAX': {'im': 170}, 
    'TAY': {'im': 168}, 
    'TSX': {'im': 186}, 
    'TXA': {'im': 138}, 
    'TXS': {'im': 154}, 
    'TYA': {'im': 152},
    'XAA': {'im': 139},
    'XAS': {'ay': 155}
}
